File: archive/compile_itunes_exports.py
#!/usr/bin/env python3
"""
iTunes Export Compiler
Compiles 368 iTunes library export files into a single CSV
"""

import csv
import os
from pathlib import Path
from typing import List, Dict, Set
import sys

def find_all_export_files(base_dir: str) -> List[Path]:
    """Find all iTunes export files across all folders"""
    base_path = Path(base_dir)
    files = []

    # Files to exclude (script outputs)
    exclude_files = {'validation_report.txt', 'compiled_itunes_library.csv',
                     'compile_itunes_exports.py'}

    # Find .txt files
    txt_files = base_path.glob("**/*.txt")
    files.extend([f for f in txt_files if f.name not in exclude_files])

    # Find Library.export* files
    files.extend(base_path.glob("**/Library.export*"))

    return sorted(files)

def detect_header_pattern(first_line: str) -> tuple[bool, int]:
    """
    Detect if the first line contains a header and where data starts.
    Returns: (has_header, num_header_fields)
    """
    fields = first_line.split('\t')

    # Standard iTunes export header fields
    standard_headers = ['Name', 'Artist', 'Composer', 'Album', 'Genre', 'Size',
                       'Time', 'Track Number', 'Year', 'Date', 'Location']

    # Check if the first few fields look like headers
    header_matches = 0
    for i, field in enumerate(fields[:15]):  # Check first 15 fields
        if field in standard_headers:
            header_matches += 1

    # If we found several standard headers, this line has headers
    if header_matches >= 5:
        # Find where headers end by looking for a common header pattern
        # Headers typically include: Name, Artist, Album, ... Location
        # Then data starts (often with a volume ID at the very end)

        # Try to find "Location" field, which is typically the last header
        try:
            location_idx = fields.index('Location')
            return True, location_idx + 1
        except ValueError:
            # If Location not found, estimate based on header matches
            # Standard format has about 25-27 fields
            return True, min(27, len(fields) // 2)

    return False, 0

def parse_export_file(filepath: Path) -> tuple[List[str], List[Dict], List[str]]:
    """
    Parse a single iTunes export file.
    Handles three cases:
    1. Files with separate header row
    2. Files with header+data on first line
    3. Files with no header (just data)
    Returns: (headers, rows, errors)
    """
    errors = []
    rows = []
    headers = []

    # Standard iTunes export headers (for files without headers)
    standard_headers = [
        'Name', 'Artist', 'Composer', 'Album', 'Grouping', 'Genre',
        'Size', 'Time', 'Disc Number', 'Disc Count', 'Track Number', 'Track Count',
        'Year', 'Date Modified', 'Date', 'Date Added',
        'Bit Rate', 'Sample Rate', 'Volume Adjustment',
        'Kind', 'Equalizer', 'Comments',
        'Play Count', 'Last Played', 'My Rating',
        'Location'
    ]

    try:
        with open(filepath, 'r', encoding='utf-8', errors='replace') as f:
            content = f.read()

        # Split into lines
        lines = content.strip().split('\n')

        if not lines:
            errors.append(f"Empty file: {filepath.name}")
            return headers, rows, errors

        # Analyze first line
        first_line = lines[0]
        has_header, num_header_fields = detect_header_pattern(first_line)

        start_line = 0

        if has_header:
            first_fields = first_line.split('\t')

            # Check if header and data are on the same line
            if len(first_fields) > num_header_fields + 5:  # Has both header and data
                # Extract header
                headers = first_fields[:num_header_fields]

                # Extract data from remaining fields
                # Last field is often a volume ID, data is between headers and vol ID
                data_fields = first_fields[num_header_fields:]

                # Create first row
                row = {}
                for i, header in enumerate(headers):
                    if i < len(data_fields):
                        row[header] = data_fields[i]
                    else:
                        row[header] = ""

                row['_source_file'] = filepath.name
                row['_line_number'] = 1
                rows.append(row)

                start_line = 1  # Continue with line 2
            else:
                # Header is on its own line
                headers = first_fields[:num_header_fields]
                start_line = 1  # Continue with line 2
        else:
            # No header detected - use standard headers
            headers = standard_headers
            start_line = 0  # Process from line 1

        # Parse remaining data rows
        for line_num, line in enumerate(lines[start_line:], start=start_line + 1):
            if not line.strip():
                continue

            fields = line.split('\t')

            # Create a dict for this row
            row = {}
            for i, header in enumerate(headers):
                if i < len(fields):
                    row[header] = fields[i]
                else:
                    row[header] = ""

            # Add source file info
            row['_source_file'] = filepath.name
            row['_line_number'] = line_num

            rows.append(row)

    except Exception as e:
        errors.append(f"Error reading {filepath.name}: {str(e)}")

    return headers, rows, errors

def get_unified_headers(all_headers: List[Set[str]]) -> List[str]:
    """
    Create a unified header list from all files
    """
    # Get all unique headers
    all_fields = set()
    for headers in all_headers:
        all_fields.update(headers)

    # Define preferred order for common fields
    preferred_order = [
        'Name', 'Artist', 'Composer', 'Album', 'Grouping', 'Genre',
        'Size', 'Time', 'Disc Number', 'Disc Count', 'Track Number', 'Track Count',
        'Year', 'Date Modified', 'Date', 'Date Added',
        'Bit Rate', 'Sample Rate', 'Volume Adjustment',
        'Kind', 'Equalizer', 'Comments',
        'Play Count', 'Last Played', 'My Rating',
        'Location'
    ]

    # Start with preferred fields that exist
    unified = [field for field in preferred_order if field in all_fields]

    # Add any remaining fields
    remaining = sorted(all_fields - set(unified) - {'_source_file', '_line_number'})
    unified.extend(remaining)

    # Add metadata fields at the end
    unified.extend(['_source_file', '_line_number'])

    return unified

def main():
    base_dir = os.path.dirname(os.path.abspath(__file__))
    output_file = os.path.join(base_dir, 'compiled_itunes_library.csv')
    validation_report = os.path.join(base_dir, 'validation_report.txt')

    print("iTunes Export Compiler")
    print("=" * 60)

    # Find all files
    print(f"\nSearching for export files in: {base_dir}")
    export_files = find_all_export_files(base_dir)
    print(f"Found {len(export_files)} export files")

    # Parse all files
    all_rows = []
    all_headers_sets = []
    all_errors = []
    file_formats = []  # Track format detection for each file
    stats = {
        'files_processed': 0,
        'files_with_errors': 0,
        'total_rows': 0,
        'empty_files': 0,
        'header_plus_data': 0,
        'separate_header': 0,
        'no_header': 0,
        'duplicates_removed': 0,
        'unique_songs': 0
    }

    print("\nParsing files...")
    for i, filepath in enumerate(export_files, 1):
        if i % 50 == 0:
            print(f"  Processed {i}/{len(export_files)} files...")

        # Read first line to detect format
        try:
            with open(filepath, 'r', encoding='utf-8', errors='replace') as f:
                first_line = f.readline().strip()

            has_header, num_fields = detect_header_pattern(first_line)
            field_count = len(first_line.split('\t'))

            if has_header and field_count > num_fields + 5:
                format_type = "header+data"
                stats['header_plus_data'] += 1
            elif has_header:
                format_type = "separate_header"
                stats['separate_header'] += 1
            else:
                format_type = "no_header"
                stats['no_header'] += 1

            file_formats.append((filepath.name, format_type, field_count))
        except:
            format_type = "unknown"
            file_formats.append((filepath.name, format_type, 0))

        headers, rows, errors = parse_export_file(filepath)

        stats['files_processed'] += 1

        if errors:
            stats['files_with_errors'] += 1
            all_errors.extend(errors)

        if not rows:
            stats['empty_files'] += 1
        else:
            all_headers_sets.append(set(headers))
            all_rows.extend(rows)
            stats['total_rows'] += len(rows)

    print(f"  Processed {len(export_files)}/{len(export_files)} files")

    # Create unified headers
    unified_headers = get_unified_headers(all_headers_sets)

    # Remove exact duplicates (excluding metadata fields)
    print(f"\nChecking for duplicates...")
    original_count = len(all_rows)

    # Create a set to track unique rows (based on content, not metadata)
    seen = set()
    deduped_rows = []
    duplicates_found = []

    for row in all_rows:
        # Create a tuple of all values except metadata fields
        content_tuple = tuple(
            row.get(field, '') for field in unified_headers
            if field not in ['_source_file', '_line_number']
        )

        if content_tuple in seen:
            # This is a duplicate
            duplicates_found.append({
                'name': row.get('Name', 'Unknown'),
                'artist': row.get('Artist', 'Unknown'),
                'source': row.get('_source_file', 'Unknown')
            })
        else:
            seen.add(content_tuple)
            deduped_rows.append(row)

    stats['duplicates_removed'] = original_count - len(deduped_rows)
    stats['unique_songs'] = len(deduped_rows)

    print(f"  Found {stats['duplicates_removed']} duplicate rows")
    print(f"  Keeping {stats['unique_songs']} unique songs")

    # Write CSV
    print(f"\nWriting compiled CSV to: {output_file}")
    with open(output_file, 'w', newline='', encoding='utf-8') as f:
        writer = csv.DictWriter(f, fieldnames=unified_headers, extrasaction='ignore')
        writer.writeheader()
        for row in deduped_rows:
            writer.writerow(row)

    # Write validation report
    print(f"Writing validation report to: {validation_report}")
    with open(validation_report, 'w', encoding='utf-8') as f:
        f.write("iTunes Export Compilation - Validation Report\n")
        f.write("=" * 60 + "\n\n")

        f.write("STATISTICS\n")
        f.write("-" * 60 + "\n")
        f.write(f"Files found: {len(export_files)}\n")
        f.write(f"Files processed: {stats['files_processed']}\n")
        f.write(f"Files with errors: {stats['files_with_errors']}\n")
        f.write(f"Empty files: {stats['empty_files']}\n")
        f.write(f"Total rows parsed: {stats['total_rows']}\n")
        f.write(f"Duplicate rows removed: {stats['duplicates_removed']}\n")
        f.write(f"Unique songs in CSV: {stats['unique_songs']}\n")
        f.write(f"Unique fields found: {len(unified_headers)}\n")
        f.write("\n")

        f.write("FILE FORMAT DETECTION\n")
        f.write("-" * 60 + "\n")
        f.write(f"Files with header+data on first line: {stats['header_plus_data']}\n")
        f.write(f"Files with separate header row: {stats['separate_header']}\n")
        f.write(f"Files with no header (data only): {stats['no_header']}\n")
        f.write("\n")

        f.write("FIELD NAMES\n")
        f.write("-" * 60 + "\n")
        for field in unified_headers:
            f.write(f"  - {field}\n")
        f.write("\n")

        f.write("FILE FORMAT DETAILS\n")
        f.write("-" * 60 + "\n")
        for filename, format_type, field_count in file_formats:
            f.write(f"  {filename:30s} | {format_type:15s} | {field_count:3d} fields\n")
        f.write("\n")

        if all_errors:
            f.write("ERRORS\n")
            f.write("-" * 60 + "\n")
            for error in all_errors:
                f.write(f"  {error}\n")
        else:
            f.write("No errors encountered!\n")

    # Print summary
    print("\n" + "=" * 60)
    print("COMPILATION COMPLETE!")
    print("=" * 60)
    print(f"Total rows parsed: {stats['total_rows']:,}")
    print(f"Duplicates removed: {stats['duplicates_removed']:,}")
    print(f"Unique songs in CSV: {stats['unique_songs']:,}")
    print(f"Files processed: {stats['files_processed']}")
    print(f"Empty files skipped: {stats['empty_files']}")
    print(f"Errors encountered: {len(all_errors)}")
    print(f"\nFile formats detected:")
    print(f"  Header+data on first line: {stats['header_plus_data']}")
    print(f"  Separate header row: {stats['separate_header']}")
    print(f"  No header (data only): {stats['no_header']}")
    print(f"\nOutput file: {output_file}")
    print(f"Validation report: {validation_report}")

    if stats['unique_songs'] > 0:
        print("\n✓ Success! Your iTunes library has been compiled.")
        return 0
    else:
        print("\n✗ No data was compiled. Check the validation report.")
        return 1

if __name__ == "__main__":
    sys.exit(main())

File: go-scripts/compile_itunes_exports.go
package main

import (
	"bufio"
	"encoding/csv"
	"fmt"
	"io/fs"
	"os"
	"path/filepath"
	"sort"
	"strconv"
	"strings"
)

var compilerExcludeFiles = map[string]struct{}{
	"validation_report.txt":       {},
	"compiled_itunes_library.csv": {},
}

var compilerHeaderProbeFields = []string{
	"Name", "Artist", "Composer", "Album", "Genre", "Size",
	"Time", "Track Number", "Year", "Date", "Location",
}

var compilerStandardHeaders = []string{
	"Name", "Artist", "Composer", "Album", "Grouping", "Genre",
	"Size", "Time", "Disc Number", "Disc Count", "Track Number", "Track Count",
	"Year", "Date Modified", "Date", "Date Added",
	"Bit Rate", "Sample Rate", "Volume Adjustment",
	"Kind", "Equalizer", "Comments",
	"Play Count", "Last Played", "My Rating",
	"Location",
}

type compileStats struct {
	FilesProcessed  int
	FilesWithErrors int
	TotalRows       int
	EmptyFiles      int
	HeaderPlusData  int
	SeparateHeader  int
	NoHeader        int
	Duplicates      int
	UniqueSongs     int
}

type formatDetail struct {
	Filename   string
	FormatType string
	FieldCount int
}

func runCompileITunesExports() {
	baseDir := Paths.ArchiveDir
	if len(os.Args) >= 3 {
		baseDir = strings.TrimSpace(os.Args[2])
	}

	outputFile := filepath.Join(baseDir, "compiled_itunes_library.csv")
	validationReport := filepath.Join(baseDir, "validation_report.txt")

	fmt.Println("iTunes Export Compiler")
	fmt.Println(strings.Repeat("=", 60))
	fmt.Printf("\nSearching for export files in: %s\n", baseDir)

	exportFiles, err := findAllExportFiles(baseDir)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error discovering export files: %v\n", err)
		os.Exit(1)
	}
	fmt.Printf("Found %d export files\n", len(exportFiles))

	allRows := make([]map[string]string, 0, 1000)
	allHeaderSets := make([]map[string]struct{}, 0, len(exportFiles))
	allErrors := make([]string, 0)
	fileFormats := make([]formatDetail, 0, len(exportFiles))
	stats := compileStats{}

	fmt.Println("\nParsing files...")
	for i, exportPath := range exportFiles {
		if (i+1)%50 == 0 {
			fmt.Printf("  Processed %d/%d files...\n", i+1, len(exportFiles))
		}

		formatType, fieldCount := detectFileFormat(exportPath)
		switch formatType {
		case "header+data":
			stats.HeaderPlusData++
		case "separate_header":
			stats.SeparateHeader++
		case "no_header":
			stats.NoHeader++
		}
		fileFormats = append(fileFormats, formatDetail{
			Filename:   filepath.Base(exportPath),
			FormatType: formatType,
			FieldCount: fieldCount,
		})

		headers, rows, parseErrors := parseExportFile(exportPath)
		stats.FilesProcessed++

		if len(parseErrors) > 0 {
			stats.FilesWithErrors++
			allErrors = append(allErrors, parseErrors...)
		}

		if len(rows) == 0 {
			stats.EmptyFiles++
			continue
		}

		headerSet := make(map[string]struct{}, len(headers))
		for _, h := range headers {
			headerSet[h] = struct{}{}
		}
		allHeaderSets = append(allHeaderSets, headerSet)
		allRows = append(allRows, rows...)
		stats.TotalRows += len(rows)
	}

	fmt.Printf("  Processed %d/%d files\n", len(exportFiles), len(exportFiles))

	unifiedHeaders := getUnifiedHeaders(allHeaderSets)

	fmt.Println("\nChecking for duplicates...")
	nonMetaHeaders := make([]string, 0, len(unifiedHeaders))
	for _, h := range unifiedHeaders {
		if h != "_source_file" && h != "_line_number" {
			nonMetaHeaders = append(nonMetaHeaders, h)
		}
	}

	seen := make(map[string]struct{}, len(allRows))
	dedupedRows := make([]map[string]string, 0, len(allRows))
	for _, row := range allRows {
		key := rowContentKey(row, nonMetaHeaders)
		if _, exists := seen[key]; exists {
			stats.Duplicates++
			continue
		}
		seen[key] = struct{}{}
		dedupedRows = append(dedupedRows, row)
	}
	stats.UniqueSongs = len(dedupedRows)

	fmt.Printf("  Found %d duplicate rows\n", stats.Duplicates)
	fmt.Printf("  Keeping %d unique songs\n", stats.UniqueSongs)

	fmt.Printf("\nWriting compiled CSV to: %s\n", outputFile)
	if err := writeCompiledCSV(outputFile, unifiedHeaders, dedupedRows); err != nil {
		fmt.Fprintf(os.Stderr, "Error writing compiled CSV: %v\n", err)
		os.Exit(1)
	}

	fmt.Printf("Writing validation report to: %s\n", validationReport)
	if err := writeValidationReport(validationReport, exportFiles, stats, unifiedHeaders, fileFormats, allErrors); err != nil {
		fmt.Fprintf(os.Stderr, "Error writing validation report: %v\n", err)
		os.Exit(1)
	}

	fmt.Println("\n" + strings.Repeat("=", 60))
	fmt.Println("COMPILATION COMPLETE!")
	fmt.Println(strings.Repeat("=", 60))
	fmt.Printf("Total rows parsed: %s\n", formatWithCommas(stats.TotalRows))
	fmt.Printf("Duplicates removed: %s\n", formatWithCommas(stats.Duplicates))
	fmt.Printf("Unique songs in CSV: %s\n", formatWithCommas(stats.UniqueSongs))
	fmt.Printf("Files processed: %d\n", stats.FilesProcessed)
	fmt.Printf("Empty files skipped: %d\n", stats.EmptyFiles)
	fmt.Printf("Errors encountered: %d\n", len(allErrors))
	fmt.Println("\nFile formats detected:")
	fmt.Printf("  Header+data on first line: %d\n", stats.HeaderPlusData)
	fmt.Printf("  Separate header row: %d\n", stats.SeparateHeader)
	fmt.Printf("  No header (data only): %d\n", stats.NoHeader)
	fmt.Printf("\nOutput file: %s\n", outputFile)
	fmt.Printf("Validation report: %s\n", validationReport)

	if stats.UniqueSongs > 0 {
		fmt.Println("\n\u2713 Success! Your iTunes library has been compiled.")
		return
	}

	fmt.Println("\n\u2717 No data was compiled. Check the validation report.")
	os.Exit(1)
}

func findAllExportFiles(baseDir string) ([]string, error) {
	files := make([]string, 0, 512)

	err := filepath.WalkDir(baseDir, func(path string, d fs.DirEntry, err error) error {
		if err != nil {
			return nil
		}
		if d.IsDir() {
			return nil
		}

		name := d.Name()
		if _, excluded := compilerExcludeFiles[name]; excluded {
			return nil
		}

		if strings.HasSuffix(name, ".txt") {
			files = append(files, path)
		}
		if strings.HasPrefix(name, "Library.export") {
			files = append(files, path)
		}

		return nil
	})
	if err != nil {
		return nil, err
	}

	sort.Strings(files)
	return files, nil
}

func detectHeaderPattern(firstLine string) (bool, int) {
	fields := strings.Split(firstLine, "\t")

	probeSet := make(map[string]struct{}, len(compilerHeaderProbeFields))
	for _, h := range compilerHeaderProbeFields {
		probeSet[h] = struct{}{}
	}

	headerMatches := 0
	maxProbe := len(fields)
	if maxProbe > 15 {
		maxProbe = 15
	}
	for i := 0; i < maxProbe; i++ {
		if _, ok := probeSet[fields[i]]; ok {
			headerMatches++
		}
	}

	if headerMatches >= 5 {
		for i, field := range fields {
			if field == "Location" {
				return true, i + 1
			}
		}
		half := len(fields) / 2
		if half > 27 {
			half = 27
		}
		return true, half
	}

	return false, 0
}

func parseExportFile(path string) ([]string, []map[string]string, []string) {
	errorsOut := make([]string, 0)
	rows := make([]map[string]string, 0)
	headers := make([]string, 0)

	raw, err := os.ReadFile(path)
	if err != nil {
		errorsOut = append(errorsOut, fmt.Sprintf("Error reading %s: %v", filepath.Base(path), err))
		return headers, rows, errorsOut
	}

	content := normalizeNewlines(string(raw))
	lines := strings.Split(strings.TrimSpace(content), "\n")
	if len(lines) == 0 {
		errorsOut = append(errorsOut, fmt.Sprintf("Empty file: %s", filepath.Base(path)))
		return headers, rows, errorsOut
	}

	firstLine := lines[0]
	hasHeader, numHeaderFields := detectHeaderPattern(firstLine)
	startLine := 0

	if hasHeader {
		firstFields := strings.Split(firstLine, "\t")
		if len(firstFields) > numHeaderFields+5 {
			headers = append(headers, firstFields[:numHeaderFields]...)

			dataFields := firstFields[numHeaderFields:]
			row := make(map[string]string, len(headers)+2)
			for i, header := range headers {
				if i < len(dataFields) {
					row[header] = dataFields[i]
				} else {
					row[header] = ""
				}
			}
			row["_source_file"] = filepath.Base(path)
			row["_line_number"] = "1"
			rows = append(rows, row)
			startLine = 1
		} else {
			headers = append(headers, firstFields[:numHeaderFields]...)
			startLine = 1
		}
	} else {
		headers = append(headers, compilerStandardHeaders...)
		startLine = 0
	}

	for i, line := range lines[startLine:] {
		lineNum := startLine + i + 1
		if strings.TrimSpace(line) == "" {
			continue
		}

		fields := strings.Split(line, "\t")
		row := make(map[string]string, len(headers)+2)
		for i, header := range headers {
			if i < len(fields) {
				row[header] = fields[i]
			} else {
				row[header] = ""
			}
		}
		row["_source_file"] = filepath.Base(path)
		row["_line_number"] = strconv.Itoa(lineNum)
		rows = append(rows, row)
	}

	return headers, rows, errorsOut
}

func getUnifiedHeaders(allHeaderSets []map[string]struct{}) []string {
	allFields := make(map[string]struct{})
	for _, set := range allHeaderSets {
		for h := range set {
			allFields[h] = struct{}{}
		}
	}

	unified := make([]string, 0, len(allFields)+2)
	for _, field := range compilerStandardHeaders {
		if _, ok := allFields[field]; ok {
			unified = append(unified, field)
		}
	}

	remaining := make([]string, 0)
	for field := range allFields {
		if field == "_source_file" || field == "_line_number" {
			continue
		}
		found := false
		for _, existing := range unified {
			if existing == field {
				found = true
				break
			}
		}
		if !found {
			remaining = append(remaining, field)
		}
	}
	sort.Strings(remaining)
	unified = append(unified, remaining...)
	unified = append(unified, "_source_file", "_line_number")

	return unified
}

func detectFileFormat(path string) (string, int) {
	raw, err := os.ReadFile(path)
	if err != nil {
		return "unknown", 0
	}

	content := normalizeNewlines(string(raw))
	firstLine := content
	if idx := strings.IndexByte(content, '\n'); idx >= 0 {
		firstLine = content[:idx]
	}
	firstLine = strings.TrimSpace(firstLine)

	hasHeader, numFields := detectHeaderPattern(firstLine)
	fieldCount := len(strings.Split(firstLine, "\t"))

	if hasHeader && fieldCount > numFields+5 {
		return "header+data", fieldCount
	}
	if hasHeader {
		return "separate_header", fieldCount
	}
	return "no_header", fieldCount
}

func rowContentKey(row map[string]string, nonMetaHeaders []string) string {
	var b strings.Builder
	for _, h := range nonMetaHeaders {
		v := row[h]
		b.WriteString(strconv.Itoa(len(v)))
		b.WriteByte(':')
		b.WriteString(v)
		b.WriteByte('\x1f')
	}
	return b.String()
}

func writeCompiledCSV(path string, headers []string, rows []map[string]string) error {
	f, err := os.Create(path)
	if err != nil {
		return err
	}
	defer f.Close()

	writer := csv.NewWriter(f)
	writer.UseCRLF = true

	if err := writer.Write(headers); err != nil {
		return err
	}

	record := make([]string, len(headers))
	for _, row := range rows {
		for i, h := range headers {
			record[i] = row[h]
		}
		if err := writer.Write(record); err != nil {
			return err
		}
	}

	writer.Flush()
	return writer.Error()
}

func writeValidationReport(
	reportPath string,
	exportFiles []string,
	stats compileStats,
	headers []string,
	formats []formatDetail,
	allErrors []string,
) error {
	f, err := os.Create(reportPath)
	if err != nil {
		return err
	}
	defer f.Close()

	w := bufio.NewWriter(f)
	defer w.Flush()

	fmt.Fprintln(w, "iTunes Export Compilation - Validation Report")
	fmt.Fprintln(w, strings.Repeat("=", 60))
	fmt.Fprintln(w)

	fmt.Fprintln(w, "STATISTICS")
	fmt.Fprintln(w, strings.Repeat("-", 60))
	fmt.Fprintf(w, "Files found: %d\n", len(exportFiles))
	fmt.Fprintf(w, "Files processed: %d\n", stats.FilesProcessed)
	fmt.Fprintf(w, "Files with errors: %d\n", stats.FilesWithErrors)
	fmt.Fprintf(w, "Empty files: %d\n", stats.EmptyFiles)
	fmt.Fprintf(w, "Total rows parsed: %d\n", stats.TotalRows)
	fmt.Fprintf(w, "Duplicate rows removed: %d\n", stats.Duplicates)
	fmt.Fprintf(w, "Unique songs in CSV: %d\n", stats.UniqueSongs)
	fmt.Fprintf(w, "Unique fields found: %d\n", len(headers))
	fmt.Fprintln(w)

	fmt.Fprintln(w, "FILE FORMAT DETECTION")
	fmt.Fprintln(w, strings.Repeat("-", 60))
	fmt.Fprintf(w, "Files with header+data on first line: %d\n", stats.HeaderPlusData)
	fmt.Fprintf(w, "Files with separate header row: %d\n", stats.SeparateHeader)
	fmt.Fprintf(w, "Files with no header (data only): %d\n", stats.NoHeader)
	fmt.Fprintln(w)

	fmt.Fprintln(w, "FIELD NAMES")
	fmt.Fprintln(w, strings.Repeat("-", 60))
	for _, field := range headers {
		fmt.Fprintf(w, "  - %s\n", field)
	}
	fmt.Fprintln(w)

	fmt.Fprintln(w, "FILE FORMAT DETAILS")
	fmt.Fprintln(w, strings.Repeat("-", 60))
	for _, fd := range formats {
		fmt.Fprintf(w, "  %-30s | %-15s | %3d fields\n", fd.Filename, fd.FormatType, fd.FieldCount)
	}
	fmt.Fprintln(w)

	if len(allErrors) > 0 {
		fmt.Fprintln(w, "ERRORS")
		fmt.Fprintln(w, strings.Repeat("-", 60))
		for _, e := range allErrors {
			fmt.Fprintf(w, "  %s\n", e)
		}
	} else {
		fmt.Fprintln(w, "No errors encountered!")
	}

	return nil
}

func normalizeNewlines(s string) string {
	s = strings.ReplaceAll(s, "\r\n", "\n")
	s = strings.ReplaceAll(s, "\r", "\n")
	return s
}

func formatWithCommas(n int) string {
	s := strconv.Itoa(n)
	if n < 1000 {
		return s
	}

	var b strings.Builder
	prefix := len(s) % 3
	if prefix == 0 {
		prefix = 3
	}
	b.WriteString(s[:prefix])
	for i := prefix; i < len(s); i += 3 {
		b.WriteByte(',')
		b.WriteString(s[i : i+3])
	}
	return b.String()
}
